### ISSUE TITLE
read_file reads only the first character of UTF-16 LE text files

### ISSUE THREAD

**Issue by bersbersbers** [tidyverse/readr#1320]
I have received an UTF-16 LE text file which looks like this:

```
$ xxd test.txt
00000000: fffe 2300 2300 4200 4c00 4f00 4300 4b00  ..#.#.B.L.O.C.K.
```

From this file, I can read only the first byte, apparently:

```
> readr::read_file("test.txt")
[1] "#"
> nchar(readr::read_file("test.txt"))
[1] 1
```

I have a rough idea what's going on (`00` being treated as an end of string?), but I wonder why `read_file` does not handle this. After all, the BOM is in place, so the encoding should be obvious.

This is a minimal example:
```
readr::read_file(as.raw(c(0xff, 0xfe, 0x23, 0x00, 0x23, 0x00)))
```

Interestingly, there is a test for this, which tests if a single (8-bit) character can be read:

https://github.com/tidyverse/readr/blob/5ac729aa9a70fef297a634d2332ff8ff1af874fb/tests/testthat/test-parsing-character.R#L62-L72

I am not sure how useful this test is in its current form, given that `FEFF` signified UTF-16 LE.

**Comment by jimhester** [comment]
I think the best way to handle UTF-16 files is to use `readr::read_file_raw()` and then pass that to `iconv()`. In any other case you are going to run into problems with the 0 bytes being interpreted as NUL string terminators by R / readr.

**Comment by jimhester** [comment]
No it is not, you used `read_lines_raw()`

**Comment by bersbersbers** [comment]
Oh. Oooh! Thank, I'll try that :)

**Comment by bersbersbers** [comment]
Thanks - that was helpful! I'll record here what I found, maybe it's helpful for anyone else, too.

```r
stringi::stri_split_lines(iconv(list(readr::read_file_raw("test.txt")), from="UTF-16"))
```
(don't forget `list`!) turned out to work. I can also use `guess_encoding()` to my advantage: 

```r
stringi::stri_split_lines(iconv(list(readr::read_file_raw("test.txt")), from=readr::guess_encoding("test.txt")$encoding[1]))
```

One thing I noticed about `readr::read_file_raw` (and `readr::read_lines_raw`, by the way) is that they don't return the BOM - so if I want to interpret the BOM myself (for endianness), or pass it on to `iconv`, then one should use `readBin`, it seems (otherwise, I somehow wouldn't trust `iconv` to get the endianness each and every time):

```r
stringi::stri_split_lines(iconv(list(readBin("test.txt", raw(), file.info("test.txt")$size)), from=readr::guess_encoding("test.txt")$encoding[1]))
```

For some reason, the snippet just above is what I, maybe naively, had expected `readr::read_lines` (and similarly, `readr::read_file`) to do.
